eventrouter/internal/checked_config.h

```diff
@@ -31,17 +35,68 @@
 #endif
 #endif
 
-/// Default to the FreeRTOS implementation. Clients may specify ER_BAREMETAL in
-/// their eventrouter_config.h and then build without FreeRTOS.
-#if !defined(ER_FREERTOS) && !defined(ER_BAREMETAL)
-#define ER_FREERTOS
-#endif
+//==============================================================================
+// Implementation Selection
+//==============================================================================
 
-/// Sanity check the implementation definitions.
+// These are the supported event router implementations. Clients select one by
+// defining ER_IMPLEMENTATION to it. For example, placing the following
+// definition in eventrouter_config.h chooses the FreeRTOS implementation:
+//
+//    #define ER_IMPLEMENTATION ER_IMPL_FREERTOS
+//
+#define ER_IMPL_FREERTOS  1
+#define ER_IMPL_POSIX     2
+#define ER_IMPL_BAREMETAL 3
+
+// In previous versions of the event router, clients specified ER_FREERTOS and
+// ER_BAREMETAL directly, instead of defining ER_IMPLEMENTATION. These checks
+// convert old choices to new ones for backward compatibility. They also make it
+// possible to specify an implementation from the command line; e.g., setting
+// `-DER_IMPLEMENTATION=ER_FREERTOS` on the command line doesn't work.
 #if defined(ER_FREERTOS) && defined(ER_BAREMETAL)
-ER_STATIC_ASSERT(0, "Only one of ER_FREERTOS and ER_BAREMETAL may be defined");
+#error "Only one of ER_FREERTOS and ER_BAREMETAL may be defined"
+#endif
+#if defined(ER_FREERTOS)
+#define ER_IMPLEMENTATION ER_IMPL_FREERTOS
+#endif
+#if defined(ER_POSIX)
+#define ER_IMPLEMENTATION ER_IMPL_POSIX
+#endif
+#if defined(ER_BAREMETAL)
+#define ER_IMPLEMENTATION ER_IMPL_BAREMETAL
+#endif
+
+// The FreeRTOS implementation is the default for historical reasons.
+#define ER_IMPLEMENTATION_DEFAULT ER_IMPL_FREERTOS
+
+// Make sure ER_IMPLEMENTATION is valid and provide a default if necessary.
+#if !defined(ER_IMPLEMENTATION)
+#define ER_IMPLEMENTATION ER_IMPLEMENTATION_DEFAULT
+#elif 1 - ER_IMPLEMENTATION - 1 == 2  // Detect "#define ER_IMPLEMENTATION".
```

Comment:
This is an interesting (?) hack to detect when someone types `#define ER_IMPLEMENTATION` with no value after it. If it's too hateful, I can try to find another approach.

Reply:
And will also detect`#define ER_IMPLEMENTATION -2` :)